test/distribution/test_distribution_methods.py

```diff
@@ -0,0 +1,142 @@
+    Gaussian,
+    MultivariateGaussian,
+    PiecewiseLinear,
+    Binned,
+    TransformedDistribution,
+)
+
+test_cases = [
+    (
+        Gaussian,
+        {
+            'mu': mx.nd.array([1000.0, -1000.0]),
+            'sigma': mx.nd.array([0.1, 1.0]),
+        },
+    ),
+    (
+        Laplace,
+        {'mu': mx.nd.array([1000.0, -1000.0]), 'b': mx.nd.array([0.1, 1.0])},
+    ),
+    (
+        StudentT,
+        {
+            'mu': mx.nd.array([1000.0, -1000.0]),
+            'sigma': mx.nd.array([1.0, 2.0]),
+            'nu': mx.nd.array([4.2, 3.0]),
+        },
+    ),
+    (
+        NegativeBinomial,
+        {'mu': mx.nd.array([1000.0, 1.0]), 'alpha': mx.nd.array([1.0, 2.0])},
+    ),
+    (
+        Uniform,
+        {
+            'low': mx.nd.array([1000.0, -1000.1]),
+            'high': mx.nd.array([2000.0, -1000.0]),
+        },
+    ),
+    (
+        Binned,
+        {
+            'bin_probs': mx.nd.array(
+                [[0, 0.3, 0.1, 0.05, 0.2, 0.1, 0.25]]
+            ).repeat(axis=0, repeats=2),
+            'bin_centers': mx.nd.array(
+                [[-5, -3, -1.2, -0.5, 0, 0.1, 0.2]]
+            ).repeat(axis=0, repeats=2),
+        },
+    ),
+]
+
+test_output = {
+    'Gaussian': {
+        'mean': mx.nd.array([1000.0, -1000.0]),
+        'stddev': mx.nd.array([0.1, 1.0]),
+        'variance': mx.nd.array([0.01, 1.0]),
+    },
+    'Laplace': {
+        'mean': mx.nd.array([1000.0, -1000.0]),
+        'stddev': mx.nd.array([0.14142136, 1.4142135]),
+        'variance': mx.nd.array([0.02, 1.9999999]),
+    },
+    'StudentT': {
+        'mean': mx.nd.array([1000.0, -1000.0]),
+        'stddev': mx.nd.array([1.3816986, 3.4641016]),
+        'variance': mx.nd.array([1.909091, 12.0]),
+    },
+    'NegativeBinomial': {
+        'mean': mx.nd.array([1000.0, 1.0]),
+        'stddev': mx.nd.array([1000.4999, 1.7320508]),
+        'variance': mx.nd.array([1.001e06, 3.000e00]),
+    },
+    'Uniform': {
+        'mean': mx.nd.array([1500.0, -1000.05]),
+        'stddev': mx.nd.array([2.8867514e02, 2.8860467e-02]),
+        'variance': mx.nd.array([8.3333336e04, 8.3292654e-04]),
+    },
+    'Binned': {
+        'mean': mx.nd.array([-0.985, -0.985]),
+        'stddev': mx.nd.array([1.377416, 1.377416]),
+        'variance': mx.nd.array([1.8972749, 1.8972749]),
+    },
+}
+
+# TODO: implement stddev methods for MultivariateGaussian and LowrankMultivariateGaussian
+DISTRIBUTIONS = [
+    Gaussian,
+    Laplace,
+    StudentT,
+    NegativeBinomial,
+    Uniform,
+    Binned,
+]
+
+
+@pytest.mark.parametrize("distr_class, params", test_cases)
+def test_means(distr_class, params) -> None:
+    distr = distr_class(**params)
+    means = distr.mean
+    distr_name = distr.__class__.__name__
+    assert means.shape == test_output[distr_name]['mean'].shape
+    # asnumpy()  needed to b/c means is all pointers to values
+    assert np.allclose(
+        means.asnumpy(), test_output[distr_name]['mean'].asnumpy(), atol=1e-2
```

Comment:
This value for `atol` seems a bit too large. For example, in some assertions we are comparing against rather small values (around `1e-4` in one case). What happens if you leave `atol` to its default value in all assertions?

Reply:
You can make them [go to 11](https://www.youtube.com/watch?v=4xgx4k83zzc) and all tests pass:
```
(gluon-ts) Lucass-MacBook:gluon-ts rlucas$ git diff --staged
diff --git a/test/distribution/test_distribution_methods.py b/test/distribution/test_distribution_methods.py
index 3c9f283..18fc80f 100644
--- a/test/distribution/test_distribution_methods.py
+++ b/test/distribution/test_distribution_methods.py
@@ -112,7 +112,7 @@ def test_means(distr_class, params) -> None:
     assert means.shape == test_output[distr_name]['mean'].shape
     # asnumpy()  needed to b/c means is all pointers to values
     assert np.allclose(
-        means.asnumpy(), test_output[distr_name]['mean'].asnumpy(), atol=1e-2
+        means.asnumpy(), test_output[distr_name]['mean'].asnumpy(), atol=1e-11
     )
 
 
@@ -125,7 +125,7 @@ def test_stdevs(distr_class, params) -> None:
     assert np.allclose(
         stddevs.asnumpy(),
         test_output[distr_name]['stddev'].asnumpy(),
-        atol=1e-2,
+        atol=1e-11,
     )
 
 
@@ -138,5 +138,5 @@ def test_variances(distr_class, params) -> None:
     assert np.allclose(
         variances.asnumpy(),
         test_output[distr_name]['variance'].asnumpy(),
-        atol=1e-2,
+        atol=1e-11,
     )
(gluon-ts) Lucass-MacBook:gluon-ts rlucas$  python -m pytest test/distribution/test_distribution_methods.py 
Setting module np/mx/python random seeds, use MXNET_MODULE_SEED=556451365 to reproduce.
================================================================================ test session starts ================================================================================
platform darwin -- Python 3.6.8, pytest-3.10.1, py-1.8.0, pluggy-0.9.0
rootdir: /Users/rlucas/gluon-ts, inifile: pytest.ini
plugins: xdist-1.27.0, timeout-1.2.1, remotedata-0.3.1, openfiles-0.3.2, forked-1.0.2, doctestplus-0.3.0, cov-2.6.1, arraydiff-0.3
collected 18 items                                                                                                                                                                  

test/distribution/test_distribution_methods.py ..................                                                                                                             [100%]

============================================================================= 18 passed in 0.16 seconds =============================================================================
(gluon-ts) Lucass-MacBook:gluon-ts rlucas$ 
```

Reply:
the `atol=1e-2` were inherited from `test/distribution/test_distribution_sampling.py` when I created a new file for testing I `cp`-ed that one.